geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/builder/TopLevelDescriptorBuilder.scala

```diff
@@ -0,0 +1,24 @@
+package org.locationtech.geomesa.utils.geotools.builder
+
+import org.geotools.feature.NameImpl
+import org.geotools.feature.`type`.AttributeDescriptorImpl
+import org.opengis.feature.`type`.{AttributeDescriptor, ComplexType, Name}
+
+import scala.collection.mutable
+
+class TopLevelDescriptorBuilder extends AbstractFeatureTypeBuilder[AttributeDescriptor] {
+  def name(n: String): DescriptorBuilder[_<:AttributeDescriptor] = {
+    _name = new NameImpl(targetNs, n)
+    `type`(n + "Type")
+  }
+
+  def end = new AttributeDescriptorImpl(makeType, _name, 1, 1, false, null)
+
+  val typeMap: scala.collection.mutable.Map[Name,ComplexType] = mutable.Map.empty
+
+  override def saveType[T <: ComplexType](t: T) = {typeMap(t.getName)=t;t}
+
+  override def getType(name:Name) = typeMap.get(name)
+
+  override def getType(name:Name, computeFn: =>ComplexType) = typeMap.getOrElseUpdate(name, computeFn)

```

Comment:
I'd suggest renaming this getTypeOrUpdate.

As a sidebar, we went through a similar naming debate regarding a map of feature types elsewhere.  The suggestion from that was to reflect the underlying methods/operations as best as possible in the function names.